src/Microsoft.OData.Client/Serialization/HttpWebRequestMessage.cs

```diff
@@ -225,12 +252,50 @@ public override string GetHeader(string headerName)
         /// <param name="headerValue">Value of the header.</param>
         public override void SetHeader(string headerName, string headerValue)
         {
-#if DEBUG
-            // Only content length header is set after firing SendingRequest2 event
-            Debug.Assert(!this.sendingRequest2Fired || headerName == XmlConstants.HttpContentLength, "!this.sendingRequest2Fired || headerName == XmlConstants.HttpContentLength");
-#endif
-            Util.CheckArgumentNullAndEmpty(headerName, "headerName");
-            HttpWebRequestMessage.SetHeaderValue(this.httpRequest, headerName, headerValue);
+            if (_contentHeaderNames.Contains(headerName))
+            {
+                // If this is a "Content" header then we cache the value (due

```

Comment:
Follow up to our previous conversation, you may find a more complete list of the content headers here https://docs.microsoft.com/en-us/dotnet/api/system.net.http.headers.httpcontentheaders?view=net-5.0#properties
I'm not 100% sure whether you need to cater for all them. I just thought you should be aware